pkg/serving/v1/apply.go

```diff
@@ -0,0 +1,293 @@
+package v1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/jsonmergepatch"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
+
+	"knative.dev/client/pkg/util"
+)
+
+// Copyright © 2020 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Helper methods supporting Apply()
+
+// patchStrategy is the patch strategy to us when sending an apply patch
+// This is fixed now to a 'normal' JSON patch, as a strategic merge patch is not supported
+// (yet) for a custom resource definition.
+var patchStrategy = types.MergePatchType
+
+func (cl *knServingClient) patch(modifiedService *servingv1.Service, currentService *servingv1.Service, uOriginalService []byte) (bool, error) {
+	uModifiedService, err := getModifiedConfiguration(modifiedService, true)
+	if err != nil {
+		return false, err
+	}
+	hasChanged, err := cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	for i := 1; i <= 5 && apierrors.IsConflict(err); i++ {
+		if i > 1 {
+			time.Sleep(1 * time.Second)
+		}
+		currentService, err = cl.GetService(currentService.Name)
+		if err != nil {
+			return false, err
+		}
+		hasChanged, err = cl.patchSimple(currentService, uModifiedService, uOriginalService)
+	}
+
+	// TODO: Introduce a "force" parameter that would do the same as `kn service create --force`
```

Comment:
Commented code...

Reply:
I'm not sure how to deal with this, as there is still a way how we can improve this initial apply implementation that has some limitations. But only then a --force option can be implemented, which is not possible right now. Before forgetting this I think its fair to keep some comments, although Im not a fan of TODO comments in general, but in this case it might be justified as the implementation is blocked by the way how the merge is done. See also my general comments above about the merge algorithm chosen.